pkg/controller/service_control.go

```diff
@@ -83,6 +84,26 @@ func (sc *realServiceControl) UpdateService(tc *v1alpha1.TidbCluster, svc *corev
 		} else {
 			svc = updated.DeepCopy()
 			svc.Spec = *svcSpec
+
+			existingSvc := updated
+			desiredSvc := svc
+			ports := existingSvc.Spec.Ports
+
+			// If the existed service and the desired service is NodePort or LoadBalancerType, we should keep the nodePort unchanged.
+			if (serviceType == corev1.ServiceTypeNodePort || serviceType == corev1.ServiceTypeLoadBalancer) &&
+				(desiredSvc.Spec.Type == corev1.ServiceTypeNodePort || desiredSvc.Spec.Type == corev1.ServiceTypeLoadBalancer) {
+				for i, dport := range existingSvc.Spec.Ports {
+					for _, eport := range ports {
+						// Because the portName could be edited,
+						// we use Port number to link the desired Service Port and the existed Service Port in the nested loop
+						if dport.Port == eport.Port && dport.Protocol == eport.Protocol {
+							dport.NodePort = eport.NodePort
+							existingSvc.Spec.Ports[i] = dport
+							break
+						}
+					}
+				}
+			}
```

Comment:
why the node port may change? because we empty the `NodePort` field each time? what's your solution to fix this?

`UpdateService` is a generic function to update service and resolve conflicts if necessary. Trying to implement this logic in this function is strange. I think we should generate correct desired service specs before calling the API update. And generating the correct desired service can help avoid unnecessary API updates (skip if the desired spec matches existing spec).

Reply:
This sounds make sense, updated.